patroni/config_generator.py

```diff
@@ -0,0 +1,466 @@
+            list(ConfigHandler.CMDLINE_OPTIONS.keys())
+
+    def _get_bin_dir_from_running_instance(self) -> str:
+        """Define the directory postgres binaries reside using postmaster's pid executable.
+
+        :param data_dir: the PostgreSQL data directory to search for postmaster.pid file in.
+
+        :returns: path to the PostgreSQL binaries directory
+
+        :raises:
+            :class:`PatroniException`: if:
+
+                * pid could not be obtained from the `postmaster.pid` file; or
+                * :exc:`OSError` occured during `postmaster.pid` file handling; or
+                * the obtained postmaster pid doesn't exist.
+        """
+        postmaster_pid = None
+        data_dir = self.config['postgresql']['data_dir']
+        try:
+            with open(f"{data_dir}/postmaster.pid", 'r') as pid_file:
+                postmaster_pid = pid_file.readline()
+                if not postmaster_pid:
+                    raise PatroniException('Failed to obtain postmaster pid from postmaster.pid file')
+                postmaster_pid = int(postmaster_pid.strip())
+        except OSError as err:
+            raise PatroniException(f'Error while reading postmaster.pid file: {err}')
+        try:
+            return os.path.dirname(psutil.Process(postmaster_pid).exe())
+        except psutil.NoSuchProcess:
+            raise PatroniException("Obtained postmaster pid doesn't exist.")
+
+    @contextmanager
+    def _get_connection_cursor(self) -> Iterator[Union['cursor', 'Cursor[Any]']]:
+        """Get cursor for the PG connection established based on the stored information.
+
+        :raises:
+            :class:`PatroniException`: if :exc:`psycopg.Error` occured.
+        """
+        try:
+            conn = psycopg.connect(dsn=self.dsn,
+                                   password=self.config['postgresql']['authentication']['superuser']['password'])
+            with conn.cursor() as cur:
+                yield cur
+            conn.close()
+        except psycopg.Error as e:
+            raise PatroniException(f'Failed to establish PostgreSQL connection: {e}')
+
+    def _is_superuser(self, cur: Union['cursor', 'Cursor[Any]'], username: str) -> bool:
+        """Check if the user has superuser privilege.
+
+        :param cur: connection cursor to use for the check.
+        :param username: username to check.
+        """
+        cur.execute("SELECT 1 FROM pg_roles WHERE rolname=%s AND rolsuper", (username,))
+        return cur.rowcount == 1
+
+    def _set_pg_params(self, cur: Union['cursor', 'Cursor[Any]']) -> None:
+        """Extend :attr:`~RunningClusterConfigGenerator.config` with the actual PG GUCs values.
+
+        THe following GUC values are set:
+
+            * Non-internal having configuration file, postmaster command line or environment variable
+              as a source.
+            * List of the always required parameters (see :func:`~RunningClusterConfigGenerator._required_pg_params`)
```

Comment:
```suggestion
            * Non-internal having configuration file, postmaster command line or environment variable
              as a source.

            * List of the always required parameters (see :func:`~RunningClusterConfigGenerator._required_pg_params`)
```

Do it doesn't merge when building with sphinx.